File: SinglyLL.c
#include<stdio.h>
#include<stdlib.h>

struct node
{
	int data;
	struct node *next;
};
typedef struct node NODE;
typedef struct node * PNODE;
typedef struct node ** PPNODE;

void InsertFirst(PPNODE head , int no)
{
	PNODE newn = NULL;
	
	newn =(PNODE)malloc(sizeof(NODE));
	
	newn->data = no;
	newn->next = NULL;

   if(*head == NULL)
   {
	   *head = newn;
   }	   
   else
   {
	   newn->next = *head;
	   *head = newn;
   }
}

void InsertLast(PPNODE head , int no)
{
	PNODE newn = NULL;
	
	newn =(PNODE)malloc(sizeof(NODE));
	
	newn->data = no;
	newn->next = NULL;

   if(*head == NULL)
   {
	   *head = newn;
   }	   
   else
   {
	   PNODE temp = *head;
	   
	   while(temp->next != NULL)
	   {
		   temp = temp->next;
	   }
	   temp->next = newn;
	   newn = temp;
   }
}

void DeleteFirst(PPNODE head)
{
	PNODE temp = NULL;
	
	if(*head == NULL)
	{
		return;
	}
	else
	{
		temp = *head;
		*head = temp->next;
		 free(temp);
	}
}

void DeleteLast(PPNODE head)
{
	PNODE temp = NULL;
	
	if(*head == NULL)
	{
		return;
	}
	else
	{
		 temp = *head;
		 
		 while(temp->next->next != NULL)
		 {
			 temp = temp->next;
		 }
		 free(temp->next);
		 temp->next = NULL;
	}	
		
}

void Display(PNODE head)
{
	PNODE temp = head;
	do
	{
		printf("|  %d  |->",temp->data);
		temp = temp->next;
	}while(temp != NULL);
	printf("NULL\n");
}
int Count(PNODE head)
{
	PNODE temp = head;
	int iCnt = 0;
	do
	{
		iCnt++;
		temp = temp->next;
	}while(temp != NULL);
	
	return iCnt;
}

void InsertAtPos(PPNODE head , int no ,int Pos)
{
	int size = 0;
	size = Count(*head);
	PNODE temp = *head;
	
	if((Pos<1)  || (Pos>(size + 1)))
	{
		return;
	}
	if(Pos == 1)
	{
		InsertFirst(head,no);
	}
	else if(Pos == (size + 1))
	{
		InsertLast(head,no);
	}
	else
	{
		PNODE newn = NULL;
		
		newn = (PNODE)malloc(sizeof(NODE));
		
		newn->data=no;
		newn->next=NULL;
		
		temp = *head;
		int iCnt = 0;
		
		for(iCnt = 1 ; iCnt < size - 1 ; iCnt++)
		{
			temp = temp->next;
		}
		newn->next=temp->next;
		temp->next=newn;
	}
}

void DeleteAtPos(PPNODE head , int no ,int Pos)
{
	int size = 0;
	size = Count(*head);
	PNODE temp = *head;
	PNODE template = NULL;
	
	if((Pos<1)  || (Pos>(size + 1)))
	{
		return;
	}
	if(Pos == 1)
	{
		DeleteFirst(head);
	}
	else if(Pos == (size))
	{
		DeleteLast(head);
	}
	else
	{
		
		temp = *head;
		int iCnt = 0;
		
		for(iCnt = 1 ; iCnt < Pos - 1 ; iCnt++)
		{
			temp = temp->next;
		}
		 template = temp->next;
		 temp->next = temp->next->next->next;
		free(template);
	}
}

int main()
{
	PNODE first = NULL;
	int iRet = 0;
	
	InsertFirst(&first , 11 );
	InsertFirst(&first , 21 );
	InsertFirst(&first , 31 );
	InsertFirst(&first , 41 );
	Display(first);
	iRet = Count(first);
	printf("Number of Elelments are:%d\n",iRet);
	
	InsertFirst(&first , 51 );
	Display(first);
	iRet = Count(first);
	printf("Number of Elelments are:%d\n",iRet);
	
	InsertLast(&first , 61 );
	Display(first);
	iRet = Count(first);
	printf("Number of Elelments are:%d\n",iRet);
	
	DeleteFirst(&first);
	Display(first);
	iRet = Count(first);
	printf("Number of Elelments are:%d\n",iRet);
	
	DeleteLast(&first);
	Display(first);
	iRet = Count(first);
	printf("Numberssss of Elelments are:%d\n",iRet);
	
	InsertAtPos(&first , 38 , 2);
	Display(first);
	iRet = Count(first);
	printf("Numberssss of Elelments are:%d\n",iRet);
	
	DeleteAtPos(&first , 38 , 2);
	Display(first);
	iRet = Count(first);
	printf("Numberssss of Elelments are:%d\n",iRet);
	return 0;
}

File: SinglyLL.java
import java.lang.*;
import java.util.*;

class node 
{
	public int data;
	public node next;
}

class SLL 
{
	
	public node Head;
	public int Count;
	
	public SLL()
	{
		Head = null;
		Count = 0;
	}
	
	public void InsertFirst(int no)
	{
		node newn = new node();
		
		newn.data = no;
		newn.next = null;
		
		if(Head == null){
			Head = newn;
		}else{
			newn.next = Head;
			Head = newn;
		}
		Count++;
	}
	
	public void InsertLast(int no)
	{
		node newn = new node();
		
		newn.data = no;
		newn.next = null;
		
		if(Head == null){
			Head = newn;
		}else{
			node temp = Head;
			
			while(temp.next != null){
				temp = temp.next;
			}
			temp.next = newn;
		}
		Count++;
	}
	public void Display()
	{
		node temp = Head;
		
		System.out.println();
		System.out.println("Elements of Linked list are : ");
		while(temp != null){
			System.out.print("|"+temp.data+"|->");
			temp = temp.next;
		}
		System.out.println("NULL");
	}
	public void InsertAtPos(int no, int pos)
	{
		if((pos < 1) || (pos > 1+Count )){
			return;
		}
		if(pos == 1){
			InsertFirst(no);
		}
		if(pos == Count+1){
			InsertLast(no);
		}else{
			node newn = new node();
			
			newn.data = no;
			newn.next = null;
			
			node temp = Head;
			
			for(int i = 1 ; i < pos - 1 ; i++){
				temp = temp.next;
			}
			
			newn.next = temp.next;
			temp.next = newn;
			
			Count++;
		}
	}
	public void DeleteFirst()
	{
		if(Count == 0){
			return;
		}else if(Count == 1){
			Head = null;
		}else{
			node temp = Head;
			
			Head = Head.next;
			temp = null;
		}
		Count--;
	}
	public void DeleteLast()
	{
		if(Count == 0){
			return;
		}else if(Count == 1){
			Head = null;
		}else{
			node temp = Head;
			while(temp.next.next != null){
				temp = temp.next;
			}
			temp.next = null;
		}
		Count--;
	}
	public int CountNode()
	{
		return Count;
	}
	public void DeleteAtPos(int Pos)
	{
		if( (Pos < 1) || (Pos > Count) ){
			return;
		}
	    if(Pos == 1){
			DeleteFirst();
		}else if(Pos == Count){
			DeleteLast();
		}else{
			node temp = Head;
			
			for(int i = 1 ; i<Pos-1 ; i++){
				temp = temp.next;
			}
			temp.next = temp.next.next;
			
			Count--;
		}
	}
}

class SinglyLL
{
	public static void main(String args[])
	{
		int iRet = 0;
		
		SLL obj = new SLL();
		
		obj.InsertFirst(8);
		obj.InsertFirst(7);
		obj.InsertFirst(6);
		obj.InsertLast(9);
		obj.InsertLast(10);
		obj.Display();
		iRet = obj.CountNode();
		System.out.println("Number of Elements are : "+obj.CountNode());
		
		obj.InsertAtPos(55,3);
		obj.Display();
		iRet = obj.CountNode();
		System.out.println("Number of Elements are: "+obj.CountNode());
		
		obj.DeleteFirst();
		obj.Display();
		iRet = obj.CountNode();
		System.out.println("Number of Elements are : "+obj.CountNode());
		
		obj.DeleteLast();
		obj.Display();
		iRet = obj.CountNode();
		System.out.println("Number of Elements are : "+obj.CountNode());
		
		obj.DeleteAtPos(3);
		obj.Display();
		iRet = obj.CountNode();
		System.out.println("Number of Elements are : "+obj.CountNode());
	}
}